File: test_src/main_subscribe.py
from pydbus import SessionBus
from gi.repository import GLib

def on_steering_changed(newSteering):
    print(f"Steering updated to: {newSteering}")

def on_speed_changed(newSpeed):
    print(f"Speed updated to: {newSpeed}")

if __name__ == "__main__":
    bus = SessionBus()
    proxy = bus.get("com.team2.VehicleControl")
    proxy.SteeringChanged.connect(on_steering_changed)
    proxy.SpeedChanged.connect(on_speed_changed)

    loop = GLib.MainLoop()
    loop.run()

File: test_src/main_subscribe.cpp
#include <dbus-c++-1/dbus-c++/dbus.h>
#include <iostream>
#include "vehicle_control_proxy.h"

class VehicleControlProxy : public com::team2::VehicleControl_proxy,
                            public DBus::IntrospectableProxy,
                            public DBus::ObjectProxy {
public:
    VehicleControlProxy(DBus::Connection& connection, const char* path, const char* name)
        : DBus::ObjectProxy(connection, path, name) {}

    virtual void SteeringChanged(const int32_t& newSteering) {
        std::cout << "Steering updated to: " << newSteering << std::endl;
    }

    virtual void SpeedChanged(const int32_t& newSpeed) {
        std::cout << "Speed updated to: " << newSpeed << std::endl;
    }
};

int main() {
    DBus::BusDispatcher dispatcher;
    DBus::default_dispatcher = &dispatcher;
    DBus::Connection conn = DBus::Connection::SessionBus();

    VehicleControlProxy proxy(conn, "/com/team2/VehicleControl", "com.team2.VehicleControl");

    dispatcher.enter();
    return 0;
}
